core/src/com/google/inject/internal/ProviderMethod.java

```diff
@@ -237,31 +237,33 @@ public int hashCode() {
 
   /*if[AOP]*/
   /**
-   * A {@link ProviderMethod} implementation that uses {@link net.sf.cglib.reflect.FastClass#invoke}
-   * to invoke the provider method.
+   * A {@link ProviderMethod} implementation that uses bytecode generation to invoke the provider
+   * method.
    */
   private static final class FastClassProviderMethod<T> extends ProviderMethod<T> {
-    final net.sf.cglib.reflect.FastClass fastClass;
-    final int methodIndex;
+    final BiFunction<Object, Object[], Object> fastMethod;
 
     FastClassProviderMethod(
         Key<T> key,
-        net.sf.cglib.reflect.FastClass fc,
         Method method,
         Object instance,
         ImmutableSet<Dependency<?>> dependencies,
         Class<? extends Annotation> scopeAnnotation,
-        Annotation annotation) {
+        Annotation annotation,
+        BiFunction<Object, Object[], Object> fastMethod) {
       super(key, method, instance, dependencies, scopeAnnotation, annotation);
-      this.fastClass = fc;
-      this.methodIndex = fc.getMethod(method).getIndex();
+      this.fastMethod = fastMethod;
     }
 
     @SuppressWarnings("unchecked")
     @Override
     public T doProvision(Object[] parameters)
         throws IllegalAccessException, InvocationTargetException {
-      return (T) fastClass.invoke(methodIndex, instance, parameters);
+      try {
+        return (T) fastMethod.apply(instance, parameters);
+      } catch (Throwable e) {
+        throw new InvocationTargetException(e);
```

Comment:
should the the FastClass/Enhancer code prewrap in InvocationTargetExceptions?

otherwise it looks like all callers need to remember to do that, and if they don't then (i'm assuming) checked exceptions from the user code can slip out?  unless it's throwing the user-errors inside of some kind of runtime exception (to make it work w/ BiFunction, in which case do we need to unwrap from that?)

(also it looks like IllegalAccessException isn't possible anymore?)

same comment throughout where we do this.

Reply:
Enhanced classes don't pre-wrap exceptions with `InvocationTargetException` because they need to match the behaviour of the original type as closely as possible. To to keep class generation simple, fast-classes also don't do any pre-wrapping. In other words the fast-class behaves as if we were calling the original class directly (from an exception PoV).

Instead we apply `InvocationTargetException` to the few fast-constructor/fast-method places that need it to match JDK reflection behaviour. (This behaviour is checked by some internal tests.) This is also closer to the calling code that unwraps the invocation exception and reports the original cause, rather than have it hidden inside the generated code.

Wrt. checked exceptions from user code: those are propagated from the BiFunction by using a common trick to get the Java compiler to treat them as unchecked (all exceptions are unchecked from the perspective of the JVM.) This means the generated code doesn't need to do any wrapping itself, only the couple of places that need to match JDK reflection behaviour.

Also wrt. IllegalAccessException - that doesn't apply to the 'fast-class' case because we only use 'fast' invocation when we know we have enough access to benefit from it, and the generated code deliberately gives enough access to Guice to use the generated invoker.

Reply:
added a bit of javadoc in 1386c48 and removed IllegalAccessException for fast-invocation